onnxruntime/core/providers/cpu/tensor/upsample.cc

```diff
@@ -227,27 +254,71 @@ Status UpsampleNearest(const T* input,
+                                    const TensorShape& input_shape, const TensorShape& output_shape,
+                                    bool is_resize) {
+  if (!input || !output) {
     return Status(ONNXRUNTIME, FAIL,
-                  is_resize ? "Resize: input / output value is nullptr"
-                            : "Upsample: input / output value is nullptr");
-  if (input_shape.NumDimensions() != output_shape.NumDimensions())
+                  is_resize ? "Resize: input/output value is nullptr"
+                            : "Upsample: input/output value is nullptr");
+  }
+
+  if (input_shape.NumDimensions() != output_shape.NumDimensions()) {
     return Status(ONNXRUNTIME, FAIL,
                   is_resize ? "Resize: input/output value's dimension mismatch"
                             : "Upsample: input/output value's dimension mismatch");
+  }
+
+  if (input_shape.NumDimensions() == 0) {
+    return Status(common::ONNXRUNTIME, common::INVALID_ARGUMENT,
+                  is_resize ? "Resize: input shape needs to be at least a single dimension"
+                            : "Upsample: input shape needs to be at least a single dimension.");
+  }
+
+  return Status::OK();
+}
+
+template <typename T>
+static Status UpsampleNearest(const T* input,
+                              T* output,
+                              const TensorShape& input_shape,
+                              const TensorShape& output_shape,
+                              const vector<float>& scales,
+                              const vector<float>& roi,
+                              bool is_resize,
+                              bool extrapolation_enabled,
+                              T extrapolation_value,
+                              bool use_nearest2x_optimization,
+                              const GetOriginalCoordinateFunc& get_original_coordinate,
+                              const GetNearestPixelFunc& get_nearest_pixel) {
+  ORT_RETURN_IF_ERROR(ValidateUpsampleInput(input, output, input_shape, output_shape, is_resize));
+
+  // special case with fast path
+  if (use_nearest2x_optimization && input_shape.NumDimensions() == 4 &&
+      scales[0] == 1 && scales[1] == 1 && scales[2] == 2 && scales[3] == 2) {
+    UpsampleNearest2x<T>(input_shape[0], input_shape[1], input_shape[2], input_shape[3], input, output);
+    return Status::OK();
+  }
+
+  return UpsampleNearestImpl(input, output, input_shape, output_shape, scales, roi,
+                             extrapolation_enabled, extrapolation_value,
+                             get_original_coordinate, get_nearest_pixel);
+}
+
+/*
+
+// This is a generic upsample in linear mode for N-D tensor.
+// But what's the correct behavior for linear mode is not clear right now.
+// this function is not enabled yet.
+// this function is not tested for opset 11 changes yet
+
+static Status UpsampleLinearImpl(const std::function<void(size_t, size_t, float)>& apply,
```

Comment:
maybe keep it uncommented out? it's easier to overlook commented out code when refactoring

Reply:
with the refactoring not commenting it out leads to a build warning/failure. 

given it has been years without this being enabled I'm not sure we need it. at least it's easier to find if it's here and commented out.